### PR TITLE
feat: archive template versions to hide them from the ui

### LINKED ISSUE #9478: Delete/purge old template versions
A user has requested to add the ability to delete or purge old versions of a template.

## Motivation 
Sometimes an old version of the template can be unusable due to many factors such as the docker image it refers doesn't exist anymore or have been renamed etc.

## Solution
In such cases, when the old versions are practically unusable it's good to allow deleting them from the CLI and UI
e.g., 

```shell
coder templates version delete VERSION_NAME
```

Originally requested in: https://github.com/coder/coder/issues/8307#issuecomment-1664277865

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

Allows "archiving" template versions to clean them up from the UX. By default archived are not listed when calling `/versions` for a template. This action is reversible. Archived versions cannot be promoted or used in workspace builds.

Template versions that are in use or the current active template version cannot be archived.

Closes https://github.com/coder/coder/issues/9478

# Cli Commands

## Archive a template

```shell
# Archive all failed template versions for the template 'docker'
$ coder templates archive docker
# Archive all unused template versions
$ coder templates archive docker --all
```

Screenshot from 2023-10-09 18-21-59

## Archive a template version

Archive/unarchive specific versions.

```shell
# Archive
$ coder templates versions archive docker strange_khorana9
Version strange_khorana9 archived at Oct  9 18:22:41


# Unarchive
$ coder templates versions unarchive docker strange_khorana9
Version strange_khorana9 unarchived at Oct  9 18:22:41
```

# Future Work

- Allow the UI to see archived versions with some toggle
- Allow UI to archive/unarchive versions

### CODE REVIEW

File: coderd/templateversions.go [hunk 1]
```diff
@@ -991,6 +1012,173 @@ func (api *API) previousTemplateVersionByOrganizationTemplateAndName(rw http.Res
...
+		}
+		if templateVersion.Archived == archive {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: fmt.Sprintf("Template version already %s", verb),
+			})
+			return
+		}
+
+		if !templateVersion.TemplateID.Valid {
+			// Maybe we should allow this?
```
Comment (Emyrk) [inline coderd/templateversions.go:1134]: I am not allowing arching orphaned versions atm, as it makes them impossible to find. Not sure what we should do with orphaned versions, but we can make this a follow up.

File: cli/templateversionarchive.go
Comment (johnstcn) [inline cli/templateversionarchive.go:1]: Can you add some tests for this CLI command?

File: cli/templateversionarchive.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+	return cmd
+}
+
+func (r *RootCmd) archiveTemplateVersions() *clibase.Cmd {
+	var all clibase.Bool
+	client := new(codersdk.Client)
+	cmd := &clibase.Cmd{
+		Use:   "archive [template-name...] ",
+		Short: "Archive unused failed template versions from a given template(s)",
```
Comment (johnstcn) [inline cli/templateversionarchive.go:104]:
```suggestion
		Short: "Archive unused or failed template versions from a given template(s)",
```

File: cli/templateversions.go [hunk 3]
```diff
@@ -29,26 +30,68 @@ func (r *RootCmd) templateVersions() *clibase.Cmd {
...
+										*v = append(*v, "Archived")
+									}
+								}
+								break
+							}
+						}
+					}
+					return next(i)
+				}
+			},
```
Comment (johnstcn) [inline cli/templateversions.go:84]: 👀 That's cool, didn't know you could do this.

File: coderd/workspaces.go [hunk 4]
```diff
@@ -370,6 +370,18 @@ func (api *API) postWorkspacesByOrganization(rw http.ResponseWriter, r *http.Req
...
+				Message: "Archived template versions cannot be used to make a workspace.",
+				Validations: []codersdk.ValidationError{
+					{
+						Field:  "template_version_id",
+						Detail: "template version archived",
+					},
+				},
+			})
+			return
+		}
```
Comment (johnstcn) [inline coderd/workspaces.go:384]: Can you add a test to assert this new behaviour?

File: coderd/templateversions_test.go [hunk 5]
```diff
@@ -1515,3 +1515,117 @@ func TestTemplateVersionParameters_Order(t *testing.T) {
 	require.Equal(t, secondParameterName, templateRichParameters[3].Name)
 	require.Equal(t, thirdParameterName, templateRichParameters[4].Name)
 }
+
+func TestTemplateArchiveVersions(t *testing.T) {
+	t.Parallel()
+
+	client := coderdtest.New(t, &coderdtest.Options{IncludeProvisionerDaemon: true})
+	user := coderdtest.CreateFirstUser(t, client)
```
Comment (johnstcn) [inline coderd/templateversions_test.go:1523]:
```suggestion
	owner := coderdtest.CreateFirstUser(t, client)
```

Also, can you create a `TemplateAdmin` user and test with that instead.

File: coderd/templateversions.go [hunk 6]
```diff
@@ -1055,6 +1243,12 @@ func (api *API) patchActiveTemplateVersion(rw http.ResponseWriter, r *http.Reque
 		})
 		return
 	}
+	if version.Archived {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "The provided template version is archived.",
+		})
+		return
+	}
```
Comment (johnstcn) [inline coderd/templateversions.go:1251]: Can you add a test to `TestPatchActiveTemplateVersion` to cover this code path?

File: coderd/templateversions.go (hunk 1 above)
Comment (johnstcn) [reply coderd/templateversions.go:1134]: I'm not sure what a user could even do with these? I think it makes sense to kick this one down the road.

File: cli/templateversions.go (hunk 3 above)
Comment (Emyrk) [reply cli/templateversions.go:84]: Yea, it's a little janky imo, but the value is tied to the option :cry:.

If we have to do this more often, I can create a helper Middleware to do this more generally.

Comment (Emyrk) [reply cli/templateversions.go:84]: This only adds the column if the user is using the default columns

File: cli/templateversionarchive.go
Comment (Emyrk) [reply cli/templateversionarchive.go:1]: Added

File: cli/templateversionarchive_test.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/coder/coder/v2/provisioner/echo"
+
+	"github.com/coder/coder/v2/codersdk"
+
+	"github.com/coder/coder/v2/testutil"
+
+	"github.com/coder/coder/v2/coderd/rbac"
+
+	"github.com/coder/coder/v2/cli/clitest"
+	"github.com/coder/coder/v2/coderd/coderdtest"
```
Comment (johnstcn) [inline cli/templateversionarchive_test.go:17]: imports

File: coderd/database/dbfake/dbfake.go [hunk 8]
```diff
@@ -5345,6 +5345,8 @@ func (q *FakeQuerier) UnarchiveTemplateVersion(_ context.Context, arg database.U
 	if err != nil {
 		return err
 	}
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
```
Comment (johnstcn) [inline coderd/database/dbfake/dbfake.go:5349]: I always forget this

**Review by johnstcn — APPROVED** [review]

**Pull request merged**